### ISSUE TITLE
Crash when initializing the SDK

### ISSUE THREAD

**Issue by lucastsantos** [prebid/prebid-mobile-android#25]
Hi,

We're building with the 0.1.0 version of the SDK and we're receiving some crash reports related to the integration.
The crash occurs when initializing the SDK in some devices.

Here's the log of this crash:

```
Fatal Exception: android.util.AndroidRuntimeException: android.content.pm.PackageManager$NameNotFoundException: com.google.android.webview
       at android.webkit.WebViewFactory.getFactoryClass(WebViewFactory.java:174)
       at android.webkit.WebViewFactory.getProvider(WebViewFactory.java:109)
       at android.webkit.WebView.getFactory(WebView.java:2194)
       at android.webkit.WebView.ensureProviderCreated(WebView.java:2189)
       at android.webkit.WebView.setOverScrollMode(WebView.java:2248)
       at android.view.View.(View.java)
       at android.view.View.(View.java)
       at android.view.ViewGroup.(ViewGroup.java)
       at android.widget.AbsoluteLayout.(AbsoluteLayout.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at org.prebid.mobile.prebidserver.internal.Settings$1.run(Settings.java:114)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:135)
       at android.app.ActivityThread.main(ActivityThread.java:5343)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:905)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:700)
Caused by android.content.pm.PackageManager$NameNotFoundException: com.google.android.webview
       at android.app.ApplicationPackageManager.getPackageInfo(ApplicationPackageManager.java:114)
       at android.webkit.WebViewFactory.getFactoryClass(WebViewFactory.java:146)
       at android.webkit.WebViewFactory.getProvider(WebViewFactory.java:109)
       at android.webkit.WebView.getFactory(WebView.java:2194)
       at android.webkit.WebView.ensureProviderCreated(WebView.java:2189)
       at android.webkit.WebView.setOverScrollMode(WebView.java:2248)
       at android.view.View.(View.java)
       at android.view.View.(View.java)
       at android.view.ViewGroup.(ViewGroup.java)
       at android.widget.AbsoluteLayout.(AbsoluteLayout.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at org.prebid.mobile.prebidserver.internal.Settings$1.run(Settings.java:114)
       at android.os.Handler.handleCallback(Handler.java:739)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:135)
       at android.app.ActivityThread.main(ActivityThread.java:5343)
       at java.lang.reflect.Method.invoke(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:372)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:905)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:700)
```

**Comment by anwzhang** [comment]
Hi @lucastsantos , for these crashes, what versions of Android and the models of devices did these happen to?

I read somewhere that for certain version of Android, if the webview is initialized too early in the app's life cycle, this could happen, but I'm not 100% sure that's the case there.

**Comment by lucastsantos** [comment]
Hi @anwzhang . The versions varies from 4.4.4 to 6.0:

![image](https://user-images.githubusercontent.com/8331460/33385188-231ad498-d50f-11e7-8bfc-065c3c89556f.png)
 
I remember MoPub having the same issue. From what I searched in that time, it appears that the WebView becomes unavailable when it's updating. This is fixed since [4.12.0](https://github.com/mopub/mopub-android-sdk/blob/master/CHANGELOG.md#version-4120-february-9-2017)

**Comment by anwzhang** [comment]
Hi @lucastsantos is this what your search result was: https://stackoverflow.com/a/29809338 ?

I could potentially add a try catch block around the webview usage in prebid server adapter settings, however, would like to pin down the root cause with you.

**Comment by lucastsantos** [comment]
@anwzhang Sorry for the delay. 
Yes, this is the explanation I saw. 
However, there's some details:
- Not all crashes are from Android 5.0. Our log contains Android 4 and 6:
![image](https://user-images.githubusercontent.com/8331460/35059613-f2b433fc-fba2-11e7-8f5e-5850d4def758.png)

- Logs are mixed between `android.content.pm.PackageManager$NameNotFoundException` and `java.lang.UnsatisfiedLinkError`. Here's the full log of this UnsatisfiedLinkError:

```
Fatal Exception: android.util.AndroidRuntimeException: java.lang.RuntimeException: Cannot load WebView
       at android.webkit.WebViewFactory.getProvider(WebViewFactory.java:107)
       at android.webkit.WebView.getFactory(WebView.java:2063)
       at android.webkit.WebView.ensureProviderCreated(WebView.java:2058)
       at android.webkit.WebView.setOverScrollMode(WebView.java:2116)
       at android.view.View.(View.java)
       at android.view.View.(View.java)
       at android.view.ViewGroup.(ViewGroup.java)
       at android.widget.AbsoluteLayout.(AbsoluteLayout.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at org.prebid.mobile.prebidserver.internal.Settings$1.run(Settings.java:114)
       at android.os.Handler.handleCallback(Handler.java:733)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5086)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by java.lang.RuntimeException: Cannot load WebView
       at com.android.org.chromium.android_webview.AwBrowserProcess.loadLibrary(AwBrowserProcess.java:31)
       at com.android.webview.chromium.WebViewChromiumFactoryProvider.(WebViewChromiumFactoryProvider.java)
       at java.lang.Class.newInstanceImpl(Class.java)
       at java.lang.Class.newInstance(Class.java:1208)
       at android.webkit.WebViewFactory.getProvider(WebViewFactory.java:102)
       at android.webkit.WebView.getFactory(WebView.java:2063)
       at android.webkit.WebView.ensureProviderCreated(WebView.java:2058)
       at android.webkit.WebView.setOverScrollMode(WebView.java:2116)
       at android.view.View.(View.java)
       at android.view.View.(View.java)
       at android.view.ViewGroup.(ViewGroup.java)
       at android.widget.AbsoluteLayout.(AbsoluteLayout.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at org.prebid.mobile.prebidserver.internal.Settings$1.run(Settings.java:114)
       at android.os.Handler.handleCallback(Handler.java:733)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5086)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by com.android.org.chromium.content.common.ProcessInitException
       at com.android.org.chromium.content.app.LibraryLoader.loadAlreadyLocked(LibraryLoader.java:135)
       at com.android.org.chromium.content.app.LibraryLoader.loadNow(LibraryLoader.java:87)
       at com.android.org.chromium.android_webview.AwBrowserProcess.loadLibrary(AwBrowserProcess.java:29)
       at com.android.webview.chromium.WebViewChromiumFactoryProvider.(WebViewChromiumFactoryProvider.java)
       at java.lang.Class.newInstanceImpl(Class.java)
       at java.lang.Class.newInstance(Class.java:1208)
       at android.webkit.WebViewFactory.getProvider(WebViewFactory.java:102)
       at android.webkit.WebView.getFactory(WebView.java:2063)
       at android.webkit.WebView.ensureProviderCreated(WebView.java:2058)
       at android.webkit.WebView.setOverScrollMode(WebView.java:2116)
       at android.view.View.(View.java)
       at android.view.View.(View.java)
       at android.view.ViewGroup.(ViewGroup.java)
       at android.widget.AbsoluteLayout.(AbsoluteLayout.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at org.prebid.mobile.prebidserver.internal.Settings$1.run(Settings.java:114)
       at android.os.Handler.handleCallback(Handler.java:733)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5086)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by java.lang.UnsatisfiedLinkError: dlopen failed: unknown reloc type 7 @ 0x56c31bfc (123758)
       at java.lang.Runtime.loadLibrary(Runtime.java:386)
       at java.lang.System.loadLibrary(System.java:526)
       at com.android.org.chromium.content.app.LibraryLoader.loadAlreadyLocked(LibraryLoader.java:123)
       at com.android.org.chromium.content.app.LibraryLoader.loadNow(LibraryLoader.java:87)
       at com.android.org.chromium.android_webview.AwBrowserProcess.loadLibrary(AwBrowserProcess.java:29)
       at com.android.webview.chromium.WebViewChromiumFactoryProvider.(WebViewChromiumFactoryProvider.java)
       at java.lang.Class.newInstanceImpl(Class.java)
       at java.lang.Class.newInstance(Class.java:1208)
       at android.webkit.WebViewFactory.getProvider(WebViewFactory.java:102)
       at android.webkit.WebView.getFactory(WebView.java:2063)
       at android.webkit.WebView.ensureProviderCreated(WebView.java:2058)
       at android.webkit.WebView.setOverScrollMode(WebView.java:2116)
       at android.view.View.(View.java)
       at android.view.View.(View.java)
       at android.view.ViewGroup.(ViewGroup.java)
       at android.widget.AbsoluteLayout.(AbsoluteLayout.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at android.webkit.WebView.(WebView.java)
       at org.prebid.mobile.prebidserver.internal.Settings$1.run(Settings.java:114)
       at android.os.Handler.handleCallback(Handler.java:733)
       at android.os.Handler.dispatchMessage(Handler.java:95)
       at android.os.Looper.loop(Looper.java:136)
       at android.app.ActivityThread.main(ActivityThread.java:5086)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:785)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:601)
       at dalvik.system.NativeStart.main(NativeStart.java)
```

**Comment by anwzhang** [comment]
Hi @lucastsantos 

This is a potential reason that caused the crash: https://corbt.com/posts/2015/09/18/mixing-32-and-64bit-dependencies-in-android.html

Let me know if it makes sense to your case.

Best,
Wei

**Comment by lucastsantos** [comment]
Hi @anwzhang 
The solution proposed in the article that you shared does not solve our problem because we do not own the native libraries in case. Note that the crash happens inside the classes related to the WebView, so we do not have control over them. 
Our app does not use WebView for anything.

As for how to reproduce the crash itself, we also were unable to do so here at our company. But since our app has a huge user base (over 1 million users per day), we received lots of error logs regarding this issue. We came to the point that we decided to not use prebid-mobile whatsoever and by doing so the problem did not happen again.

My guess is that the error happens with a combination of WebView updating + starting WebView early caused both of this crashes. However, it's practically impossible to reproduce these errors.

Is the WebView really necessary?
